### PR TITLE
Plane: tailsitter gain scailing interpolation mode

### PR DESCRIPTION
**Pull request by IamPete1**
This is some change that allow interpolation scaling to work as intended. The issue was basicaly a order of operations one. All the normal tailsitter scaling must be done before the interpolation is applied.

I have also changed the tailsitter_gscl_mask  numbers to be binary bitmasks, I think its slightly easier to understand.

There are a couple of change that might affect the other methods too. I have made some changes that put the sailing back to how it was when I first did the saciling PR. now scailing applys to tilits also and there is no less than one limit. The scaling can now also be used in combination (no idea if this is a good idea or not). I have also applied the extra elevator after the sailing.

### CODE REVIEW

**Comment by kd0aij** [comment]
Thanks for looking at this again.

I'm pretty sure the boolean value of
((tailsitter.gain_scaling_mask & TAILSITTER_GSCL_INTERP) != 0)
is the same as 
(tailsitter.gain_scaling_mask & TAILSITTER_GSCL_INTERP)
Is this a style thing? I think the second way is used more often in ArduPilot.

Was the issue you're referring to only in FW flight modes?

I don't think that it makes sense to scale motor tilts with airspeed; tilting a motor has the same effect on the thrust vector regardless of airspeed.  Does this make a difference with the Bronco? I probably wasn't looking at the same  things in my sitl testing as you are.

I'm not sure what you mean by "no less than one limit" was that in an earlier version of the code?

Perhaps it would be best to break this PR up into 3 parts which can be tested separately, a base PR which supports the bitmask and the original boost option, another for gain interpolation and a third for the  att_thr option.
That should also improve readability by encapsulating the 3 methods and making it clear how and where the scaling is calculated.

**Comment by IamPete1** [comment]
((tailsitter.gain_scaling_mask & TAILSITTER_GSCL_INTERP) != 0)  came up in the PR I did for differential spoiler mixing, I was asked to change to !=0 for some reason I cant remember.

It's a issue for all flight modes. 

Scaling tilts with airspeed is a bad idea as you say, however they should be scaled with throttle. ie the original scaling stuff. This would need a change from this PR. I was only testing my interpolation + original scaling (mask of 5). So your correct in that this might do something odd with your scaling.

The less than one limit is here 
https://github.com/kd0aij/ardupilot/blob/9c4ec7d71771e5e17916011524df3cc56a7e0b7c/ArduPlane/tailsitter.cpp#L404
it used to be constrained between 0 and max scaling.  I also removed this https://github.com/kd0aij/ardupilot/blob/9c4ec7d71771e5e17916011524df3cc56a7e0b7c/ArduPlane/tailsitter.cpp#L399
(spd_scaler >= 1.0f) and made the scailing *= so you can use them both. Maybe this is not a good idea?

3 PR's might make sense if we can make them agree somewhat so we don't end up changing a bunch of stuff back and forth. I'm happy to do the interpolation one.

I think one of the issue is that each of our method uses the original stuff in a different way.

**Comment by IamPete1** [comment]
I might have another go this eve, see if I can change this round abit so as not to break your stuff.

File: ArduPlane/tailsitter.cpp [hunk 1]
```diff
@@ -343,67 +353,62 @@ bool QuadPlane::in_tailsitter_vtol_transition(void) const
...
-            spd_scaler = constrain_float(hover_throttle / throttle, 1.0f, tailsitter.throttle_scale_max);
-        }
+    if ((temp_spd_scaler >= 1.0f) && (tailsitter.gain_scaling_mask & TAILSITTER_GSCL_BOOST) != 0) {
+        spd_scaler = MAX(1.0f, spd_scaler);
+    } else {
+        spd_scaler = temp_spd_scaler;
     }
 
-    return spd_scaler;
+    last_spd_scaler = spd_scaler;
```
Comment (IamPete1) [inline ArduPlane/tailsitter.cpp]: this should now be the only change of behavior from your orignal PR for your methods, I have put this here so the boost  throttle method is included in the filter, I think this makes sence?

**Comment by IamPete1** [comment]
hopefully this now works as it should for your stuff

File: ArduPlane/tailsitter.cpp (hunk 1 above)
Comment (kd0aij) [reply ArduPlane/tailsitter.cpp]: It looks like scaling boost is applied regardless of whether TAILSITTER_GSCL_BOOST is set
If that check is done up above, then the test here can probably just go away, I think...

Comment (IamPete1) [reply ArduPlane/tailsitter.cpp]: This was it preserves the behavior of your original PR (I think?) I'm happy with this bit however you like, it should only affect the TAILSITTER_GSCL_ATT_THR method.

**Comment by kd0aij** [comment]
stale

**Pull request closed without merging**